### PR TITLE
Issue/8536 track media error messages in db

### LINKED ISSUE #8536: Save Media error messages on CoreData.
In order to be able to track/save errors than happen on the background these need to be persisted on CoreData in order to show details to user later on.

### PR DESCRIPTION
**Pull request by SergioEstevao**
Fixes #8536 

To test:
 - Start a new post in aztec
 - Try to add/upload a media object
 - Make it fail on the middle of the process ( turn of connectivity for example)
 - See if attachment switches to the correct state
 - Check if actions on media are correct by tapping on it
 - Enable connectivity again
 - Retry the upload of media
 - Check if uploads correctly and status is updated correctly.
 - On the media library, make an upload fail also
 - Tap on media that failed, check if alert show message with the details of why the upload failed.

### CODE REVIEW

**Comment by frosty** [comment]
@SergioEstevao When the upload fails in Aztec, I still see the progress bar along the top of the media item at the % when it failed (the item does change to show the failed state though). Is this expected, or should the bar be hidden on failure?

**Comment by frosty** [comment]
The same happens for Media Library cells – they show the progress until the cell gets reused, at which point they switch to Failed.

File: MIGRATIONS.md [hunk 1]
```diff
@@ -3,6 +3,11 @@
 This file documents changes in the data model. Please explain any changes to the
 data model as well as any custom migrations.
 
+## WordPress 71
+
+- @sergioestevao 2018-01-29
+- ``Media` added `error` Transformable property that stores a NSError object tha esulted from a failed import or upload.
```
Comment (frosty) [inline MIGRATIONS.md]: Typo - 'tha esulted' -> 'that resulted'

File: WordPress/Classes/Models/Media.h [hunk 2]
```diff
@@ -45,6 +45,8 @@ typedef NS_ENUM(NSUInteger, MediaType) {
 @property (nonatomic, strong, nullable) NSString *title;
 @property (nonatomic, strong, nullable) NSString *videopressGUID;
 @property (nonatomic, strong, nullable) NSNumber *width;
+@property (nonatomic, strong, nullable) NSError *error;
```
Comment (frosty) [inline WordPress/Classes/Models/Media.h]: Should we name this `uploadError` or do you think `error` will be descriptive enough by itself?

File: WordPress/Classes/Services/MediaService.m [hunk 3]
```diff
@@ -68,10 +68,12 @@ - (Media *)createMediaWith:(id<ExportableAsset>)exportable
         [self.managedObjectContext obtainPermanentIDsForObjects:@[media] error:nil];
         [self.managedObjectContext save: nil];
     }];
-
+    NSManagedObjectID *mediaObjectID = media.objectID;
     [self.managedObjectContext performBlock:^{
         // Setup completion handlers
         void(^completionWithMedia)(Media *) = ^(Media *media) {
+            media.remoteStatus = MediaRemoteStatusLocal;
```
Comment (frosty) [inline WordPress/Classes/Services/MediaService.m]: Does this mean we'll go created -> processing -> local -> pushing -> sync?

File: WordPress/Classes/Models/Media.h (hunk 2 above)
Comment (SergioEstevao) [reply WordPress/Classes/Models/Media.h]: The error can also be a processing/importing error, so I think error is good.

File: WordPress/Classes/Services/MediaService.m (hunk 3 above)
Comment (SergioEstevao) [reply WordPress/Classes/Services/MediaService.m]: Yep, this could be useful for offline mode or in moment where we don't have connectivity.

**Comment by SergioEstevao** [comment]
@frosty 
> When the upload fails in Aztec, I still see the progress bar along the top of the media item at the % when it failed (the item does change to show the failed state though). Is this expected, or should the bar be hidden on failure?

I agree with you in both scenarios, it looks this was already like this before i develop, and strangely enough it does look we reset the progress at least visually to nil/non drawable.

Should we investigate this on a separate PR?

**Pull request merged**